actpool/queueworker.go

```diff
@@ -0,0 +1,286 @@
+		worker.addDestinationMap(act)
+	}
+
+	atomic.AddUint64(&worker.ap.gasInPool, intrinsicGas)
+
+	worker.removeEmptyAccounts()
+
+	return nil
+}
+
+func (worker *queueWorker) getConfirmedState(ctx context.Context, sender address.Address) (uint64, *big.Int, error) {
+	worker.mu.RLock()
+	queue := worker.accountActs[sender.String()]
+	worker.mu.RUnlock()
+	// account state isn't cached in the actpool
+	if queue == nil {
+		confirmedState, err := accountutil.AccountState(ctx, worker.ap.sf, sender)
+		if err != nil {
+			return 0, nil, err
+		}
+		return confirmedState.PendingNonce(), confirmedState.Balance, nil
+	}
+	return queue.AccountNonce(), queue.AccountBalance(), nil
+}
+
+func (worker *queueWorker) checkSelpWithState(act *action.SealedEnvelope, pendingNonce uint64, balance *big.Int) error {
+	if act.Nonce() < pendingNonce {
+		_actpoolMtc.WithLabelValues("nonceTooSmall").Inc()
+		return action.ErrNonceTooLow
+	}
+
+	// Nonce exceeds current range
+	if act.Nonce()-pendingNonce >= worker.ap.cfg.MaxNumActsPerAcct {
+		hash, _ := act.Hash()
+		log.L().Debug("Rejecting action because nonce is too large.",
+			log.Hex("hash", hash[:]),
+			zap.Uint64("startNonce", pendingNonce),
+			zap.Uint64("actNonce", act.Nonce()))
+		_actpoolMtc.WithLabelValues("nonceTooLarge").Inc()
+		return action.ErrNonceTooHigh
+	}
+
+	if cost, _ := act.Cost(); balance.Cmp(cost) < 0 {
+		_actpoolMtc.WithLabelValues("insufficientBalance").Inc()
+		sender := act.SenderAddress().String()
+		actHash, _ := act.Hash()
+		log.L().Info("insufficient balance for action",
+			zap.String("actionHash", hex.EncodeToString(actHash[:])),
+			zap.String("cost", cost.String()),
+			zap.String("balance", balance.String()),
+			zap.String("sender", sender),
+		)
+		return action.ErrInsufficientFunds
+	}
+	return nil
+}
+
+func (worker *queueWorker) putAction(sender string, act action.SealedEnvelope, pendingNonce uint64, confirmedBalance *big.Int) error {
+	worker.mu.RLock()
+	queue := worker.accountActs[sender]
+	worker.mu.RUnlock()
+
+	if queue == nil {
+		queue = NewActQueue(worker.ap,
+			sender,
+			pendingNonce,
+			confirmedBalance,
+			WithTimeOut(worker.ap.cfg.ActionExpiry),
+		)
+		worker.mu.Lock()
+		worker.accountActs[sender] = queue
+		worker.mu.Unlock()
+	}
+
+	if err := queue.Put(act); err != nil {
+		actHash, _ := act.Hash()
+		_actpoolMtc.WithLabelValues("failedPutActQueue").Inc()
+		log.L().Info("failed put action into ActQueue",
+			zap.String("actionHash", hex.EncodeToString(actHash[:])),
+			zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (worker *queueWorker) addDestinationMap(act action.SealedEnvelope) {
+	worker.ap.accountDesActs.mu.Lock()
+	defer worker.ap.accountDesActs.mu.Unlock()
+	destn, _ := act.Destination()
+	actHash, _ := act.Hash()
+	if desQueue := worker.ap.accountDesActs.acts[destn]; desQueue == nil {
+		worker.ap.accountDesActs.acts[destn] = make(map[hash.Hash256]action.SealedEnvelope)
+	}
+	worker.ap.accountDesActs.acts[destn][actHash] = act
```

Comment:
if this is related to ap.accountDesActs, maybe we should wrap it as a function of ap.accountDesActs.

Reply:
Sure, good suggestions